lisp/mastodon-toot.el

```diff
@@ -178,38 +204,136 @@ e.g. mastodon-toot--send -> Send."
+	 (buffer (or buffer-exists (get-buffer-create buffer-name))))
     (switch-to-buffer-other-window buffer)
-    (when (not buffer-exists)
-      (mastodon-toot--display-docs)
-      (mastodon-toot--setup-as-reply reply-to-user reply-to-id))
+    
+    ;; prevent duplication of UI stuff
+    (when (not mastodon-toot--edit-data)
+      (mastodon-toot--init-data reply-to-user)
+      (mastodon-toot--setup-as-reply reply-to-user reply-to-id)
+      (mastodon-toot--draw-ui mastodon-toot--edit-data))
     (mastodon-toot-mode t)))
 
+;;; hooks to make editing sensible
+(defun mastodon-toot--stay-in-bounds (change-start-pos change-end-pos)
+  "Use CHANGE-START-POS and CHANGE-END-POS to keep edits between the docs and the char count."
+  (let ((start-lim  (cdr (assq 'start mastodon-toot--edit-data)))
+	(end-lim  (+ 1 (cdr (assq 'end mastodon-toot--edit-data)))))
+    (message (format "%s:%s %s:%s"
+		     change-start-pos start-lim
+		     change-end-pos end-lim))
+    (cond
+     ((< change-start-pos start-lim)
+      (goto-char start-lim))
+     ((< end-lim change-start-pos)
+      (goto-char end-lim)))))
+
+(defun mastodon-toot--update-char-count (&rest _)
+  "Update the character count in MASTODON-TOOT--EDIT-DATA."
+  (let* ((raw-len (length (buffer-string)))
+	 (char-count-len (length (mastodon-toot--format-char-count
+				  (cdr (assq 'count mastodon-toot--edit-data))
+				  mastodon-toot-character-limit)))
+	 (doc-string-len (length (cdr (assq 'docs mastodon-toot--edit-data))))
+	 (final-len (- raw-len char-count-len doc-string-len))
+	 (start (cdr (assq 'start mastodon-toot--edit-data))))
+    (message (format "l:%s s:%s e:%s c:%s"
+		     final-len
+		     start
+		     (cdr (assq 'end mastodon-toot--edit-data))
+		     (cdr (assq 'count mastodon-toot--edit-data))))
+    (setf (cdr (assq 'end mastodon-toot--edit-data))
+	  (+ final-len start)
+
+	  (cdr (assq 'count mastodon-toot--edit-data))
+	  final-len)))
+
+(defun mastodon-toot--render-char-count (&rest _)
+  "Clear and render the character counter from MASTODON-TOOT--EDIT-DATA."
+  (let* ((replace-start (+ 1 (cdr (assq 'end mastodon-toot--edit-data))))
+	 (replace-end (point-max))
+	 (count (cdr (assq 'count mastodon-toot--edit-data)))
+	 (count-str (mastodon-toot--format-char-count
+		     count mastodon-toot-character-limit)))
+    (save-mark-and-excursion
+     ;; this is a "trick" from the emacs docs for deleting text
+     ;; that would otherwise be read only
+     ;; (we don't want it user editable, but we do need to update it)
+     (setq-local inhibit-read-only t)
+     (remove-text-properties replace-start replace-end '(read-only))
+     (delete-region (- replace-start 1) replace-end)
+     (goto-char replace-start)
+     (insert count-str)
+     (setq-local inhibit-read-only nil))))
+
+(defun mastodon-toot--after-change (&rest _)
+  "Dah."
+  (progn
+    (mastodon-toot--update-char-count nil)
+    (mastodon-toot--render-char-count nil)))
+
+(defun mastodon-toot--mode-hooks ()
```

Comment:
`mastodon-toot--inititalize-minor-mode` ?
That's what it is for right?  Saves you renaming in the future if/when you want to do more stuff.